### DISCUSSION TITLE
The ability to have public facing file paths that are pre-determinate

### DISCUSSION THREAD

**Discussion by dschissler** [pocketbase/pocketbase#1431] [Q&A]
I have a `static` collection that I've begun putting all of my static assets into.  In this situation I really don't need any security around the filename code being there.  It would be nice if there was an option to turn off this code suffix of the file names, or to be able to reach the files by their id, or even by a field name.

As it is now it's not quite ideal for stuffing a lot of files into it for use as static resources in a base collection that only has a single file field and nothing else.  I can also see that this is not something that was considered because in the listing the file field is without the filename when there is an entire screen for it.

@ganigeorgiev Is it always necessary to know the name of a converted filename to be able to access it?

This also creates a situation in which the API needs to be hit before knowing the location of a file.  For example a user's avatar could be something that doesn't need any security for it.  So it would be a lot easier to display a user's avatar by pointing to a determinate path of where that would be if it existed instead of needing to call the API first to get it's path.

Also I might want to have (and I do) the ability to grab a static resource like `login-image` by a field name that isn't the filename.

**Comment by ganigeorgiev** [comment]
> Is it always necessary to know the name of a converted filename to be able to access it?

Yes, unless you have your own custom static serving endpoint.

> This also creates a situation in which the API needs to be hit before knowing the location of a file. For example a user's avatar could be something that doesn't need any security for it. 

File resources are part of the record. In your case, you'll already have the user so technically you should also have access to the file name.

I'm not sure that I understand your use case and what you are trying to do. If you are working with predetermined static files wouldn't it be better to create an endpoint that serves the directory with these files instead of uploading them to a collection (eg. similar to `pb_public`)?

```go
app.OnBeforeServe().Add(func(e *core.ServeEvent) error {
    e.Router.GET("/static/*", apis.StaticDirectoryHandler(os.DirFS("./static"), false))
    return nil
})
```

The random part of the filename for now cannot be removed. Private/protected files will be handled in  the near future as part of https://github.com/pocketbase/pocketbase/issues/215 (_it is planned after the Admin UI improvements_).

**Reply by dschissler** [reply]
> I'm not sure that I understand your use case and what you are trying to do. If you are working with predetermined static files wouldn't it be better to create an endpoint that serves the directory with these files instead of uploading them to a collection (eg. similar to pb_public)?

I might put these in `pb_public` instead.  The idea is not to have such a huge binary.  I don't know why I didn't think of `pb_public` again.  Well at any rate I learned about the file upload system more than I would have.

I think though that there are ideal use cases for; 1) Golang embedded resources, 2) `pb_public` resources, 3) and files stored in a collection.  For example picking a random photo from the collection either; by 1) total records, or 2) one photo from a multi-file field.

It would be nice to have the option for no security for a collection or path, and then being able to quickly get the number of files in a field, and then from that being able to automatically know the path based upon just an index instead of the filenames.  By file name isn't always relevant or handy.

In conclusion, for now `pb_public` is better for what I'm trying to do, but I also could use a way to shuffle through a single record with a file multi-field by only knowing the total number of files that are stored there (plus other fixed path settings like collection name, and record name/id).

**Reply by dschissler** [reply]
After reading #215 I think that what is needed is a private/public system, and also some way of being able to offer how files are available, such as by index in a multi-field.  Maybe setting a file to public would automatically allow the other method of access (without using the name).  That wouldn't have a security issue because they would have already marked it as public.
